src/native/minipal/UnicodeDataGenerator/unicodedata.cs

```diff
@@ -0,0 +1,111 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Diagnostics;
+using System.Globalization;
+using System.IO;
+using System.Linq;
+
+Console.WriteLine(@"
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+//
+// THIS FILE IS GENERATED. DO NOT HAND EDIT.
+// IF YOU NEED TO UPDATE UNICODE VERSION FOLLOW THE GUIDE AT src/libraries/System.Private.CoreLib/Tools/GenUnicodeProp/Updating-Unicode-Versions.md
+//
+
+#include <minipal/utf8.h>
+
+typedef struct
+{
+  CHAR16_T code;
+  uint8_t  upperOrLower;
+  CHAR16_T opposingCode;
+} UnicodeDataRec;
+
+#define UPPER_CASE 0
+#define LOWER_CASE 1
+
+const UnicodeDataRec UnicodeData[] =
+{");
+
+string sourceFileName = args[0];
+
+int numberOfCases = 0;
+using StreamReader sourceFile = File.OpenText(sourceFileName);
+while (sourceFile.ReadLine() is string line)
+{
+    var fields = line.Split(';');
+
+    var code = int.Parse(fields[0], NumberStyles.HexNumber);
+
+    bool hasUpperCaseMapping = fields[12].Length != 0;
+    bool hasLowerCaseMapping = fields[13].Length != 0;
+
+    if (!hasLowerCaseMapping && !hasUpperCaseMapping)
+        continue;
+
+
+    int opposingCase = hasUpperCaseMapping ?
+        int.Parse(fields[12], NumberStyles.HexNumber) :
+        int.Parse(fields[13], NumberStyles.HexNumber);
+
+    // These won't fit in 16 bits - no point carrying them
+    if (code > 0xFFFF)
+        continue;
+
+    Debug.Assert(opposingCase <= 0xFFFF);
+
+    string specifier = hasUpperCaseMapping ? "LOWER_CASE" : "UPPER_CASE";
+    Console.WriteLine($"    {{ 0x{code:X}, {specifier}, 0x{opposingCase:X} }},");
+
+    numberOfCases++;
+}
+
+Console.WriteLine("};");
+
+Console.WriteLine($@"
+#define UNICODE_DATA_SIZE {numberOfCases}");
+
+Console.WriteLine(@"
+static int UnicodeDataComp(const void *opposingCode, const void *elem)
+{
+    CHAR16_T code = ((UnicodeDataRec*)elem)->code;
+
+    if (*((CHAR16_T*)opposingCode) < code)
+    {
+        return -1;
+    }
+    else if (*((CHAR16_T*)opposingCode) > code)
+    {
+        return 1;
+    }
+    else
+    {
+        return 0;
+    }
+}
+
+const CHAR16_T minipal_toupper_invariant(const CHAR16_T code)
+{
+    UnicodeDataRec *record = (UnicodeDataRec *) bsearch(&code, UnicodeData, UNICODE_DATA_SIZE,
```

Comment:
This has O(logN) complexity as before. I have a [CHM](https://cmph.sourceforge.net/chm.html) (a perfect hashing function) implementation, which increases the size of libcoreclr by ~33 KB but speeds up the lookup to O(1). Better version in my local stash, but here is an old gist: https://gist.github.com/am11/22eaa6584de55483d988d9831899bcd3.

If we want to bump the JIT vectorization limit from 0xFF to 0xFFFF for invariant upper/lowercases: https://github.com/dotnet/runtime/blob/4b87c13150990dca27497f6d37167b41562b2399/src/coreclr/jit/importervectorization.cpp#L49

and the TP regression is unacceptable, then this size tradeoff is probably an option to consider.

cc @EgorBo

Reply:
@am11 to be fair, I do not follow how JIT is related here, can you elobarate? 🙂 do you wan to extend the `ConvertToLowerCase` in JIT with Latin1? I do not see any problem with that 🙂

Reply:
> do you want to extend the `ConvertToLowerCase` in JIT with Latin1?

Yup, extending the support from 0xFF to 0xFFFF and also handling of InvariantCultureIgnoreCase (in addition to OrdinalIgnoreCase) if possible. 🙂

>  I don't think there will be any JIT TP impact from that

The current lookup with bsearch(3) is a bit slower (`O(logN)`). With the CHM approach it becomes `O(1)`.

Reply:
> > do you want to extend the `ConvertToLowerCase` in JIT with Latin1?
> 
> Yup, extending the support from 0xFF to 0xFFFF and also handling of InvariantCultureIgnoreCase (in addition to OrdinalIgnoreCase) if possible. 🙂
> 
> > I don't think there will be any JIT TP impact from that
> 
> The current lookup with bsearch(3) is a bit slower (`O(logN)`). With the CHM approach it becomes `O(1)`.

That code is executed in only a few cases and the input's length is limitted (to say, 64 bytes) so even O(n!) won't make any impact on JIT TP report 😆

Reply:
Ah, makes sense. I didn't realized optimization is enabled only for the small input sizes. 😅

Reply:
Ah, makes sense. I didn't realize optimization is enabled only for the small input sizes. 😅

Reply:
InvariantCultureIgnoreCase is implemented by ICU. You would have to hook up the JIT through CoreLib to ICU to make that work.

I think the globalization data should be provided to the JIT via JIT/EE interface. I do not think we want to be compiling a copy of the globalization data into the JIT for these sort of optimizations.

Reply:
@jkotas, I agree that we should carefully think about additional compilation units going into JIT and generally prefer the interface approach. However in this case, those particular change-case optimizations implemented in importervectorization are currently limited to 8-bits range, which can be extended to 16-bit without any additional compilation unit (at least for non-AOT).

This invariant upper/lower case data has always been linked into libclrjit.dylib via PAL on Unix and JIT has access to it without any additional cost (Windows has touppoer/tolower which support wchar). It is limited to a subset of 16-bits characters which have altering casing. It is generated using the same UnicodeData.txt which the given version of runtime libs use for System.Globalization.

Reply:
I am not worried about the extra 30kB. I am worried about having proper design for JIT/EE interface and isolation of concerns. Globalization is policy heavy component that should not be baked into the JIT. It should be exposed to the JIT via JIT/EE interface.

Reply:
Ah, I see the point. It was a side topic on my mind ever since @EgorBo made the implementation for ASCII characters, but not tied to this (mechanical) change. :)